### ISSUE TITLE
-Zdrop-tracking doesn't work if the value is borrowed

### ISSUE THREAD

**Issue by ChayimFriedman2** [rust-lang/rust#101135]
I tried this code with `-Zdrop-tracking`:

```rust
fn assert_send<T: Send>(_v: T) {}

struct NonSend(*const ());
fn use_non_send(_: &NonSend) {}

fn main() {
    assert_send(async {
        let non_send = NonSend(&());
        use_non_send(&non_send);
        drop(non_send);
        async {}.await;
    });
}
```

I expected to see this happen: this should compiler, as `-Zdrop-tracking` should see that we drop the non-`Send` type before the yield point.

Instead, this happened: this does not compile with "future cannot be sent between threads safely" pointing to the `NonSend` type.

If I comment the `use_non_send(&non_send);` line it works fine.

I don't know if this is supposed to work but this looks like a bug.

### Meta

`rustc --version --verbose`:
```
rustc 1.65.0-nightly (c07a8b4e0 2022-08-26)
binary: rustc
commit-hash: c07a8b4e09f356c7468b69c50cac7fc5b5000b8a
commit-date: 2022-08-26
host: x86_64-pc-windows-msvc
release: 1.65.0-nightly
LLVM version: 15.0.0
```

**Comment by jyn514** [comment]
cc https://github.com/rust-lang/rust/issues/97331, @eholk

I don't think this needs to block stabilization.

**Comment by eholk** [comment]
Agreed, drop tracking is still a significant improvement over the status quo and should be compatible with any fixes we do here. That said, I'd really like to fix this issue as well.

This will need some work in the MIR side of things, since that is where the more conservative treatment of borrowing comes from. On the typeck side, I think drop tracking is naturally more precise, but we've had to make it more pessimistic to agree with what MIR does.

Once the MIR is more precise, all we need to do from the typeck is basically remove anything to do with [`borrowed_temporaries`](https://github.com/rust-lang/rust/blob/98f3001eecbe4cbd091c10ffab45b4c164bb507b/compiler/rustc_typeck/src/check/generator_interior/drop_ranges/record_consumed_borrow.rs#L36).

**Comment by eholk** [comment]
We discussed this in the [wg-async triage meeting](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async/topic/Triage.20meeting.202022-09-12/near/298406210) today. Fixing this bug is something I'd like to see, but it sounds like it might be hard from a semantics standpoint. @tmandry points out that we don't know anything about about what `use_non_send` does, and that it could save a pointer that's used in unsafe code in ways that are considered sound today. So fixing this is going to require careful consideration and possibly even adjustment of unsafe code rules.

@rustbot label +AsyncAwait-Triaged

**Comment by jyn514** [comment]
@tmandry I don't understand how this would invalidate any code that was sound before; the value is dropped before the await, so any unsafe code using it afterwards is already UB.
